protodata/src/main/kotlin/io/spine/kanban/codegen/ErrorMessage.kt

```diff
@@ -0,0 +1,153 @@
+        @JvmOverloads
+        fun forRule(format: String, value: String = "", other: String = "") =
+            ErrorMessage(
+                format
+                    .replacePlaceholder(VALUE, value)
+                    .replacePlaceholder(OTHER, other)
+            )
+
+        /**
+         * Produces an error message for a composite validation rule.
+         *
+         * @param format the message format
+         * @param left the error message of the left rule
+         * @param right the error message of the right rule
+         * @param operation the operator which joins the rule conditions
+         */
+        @JvmStatic
+        @JvmOverloads
+        fun forComposite(format: String,
+                         left: ErrorMessage,
+                         right: ErrorMessage,
+                         operation: BinaryOperation = BO_UNKNOWN) =
+            ErrorMessage(
+                format
+                    .replacePlaceholder(LEFT, left.value)
+                    .replacePlaceholder(RIGHT, right.value)
+                    .replacePlaceholder(OPERATION, operation.printableString())
+            )
+    }
+
+    override fun toString() = value
+
+    /**
+     * Constructs code which creates a [ConstraintViolation] of a simple validation rule and adds it
+     * to the given mutable [violationsList].
+     */
+    fun createViolation(field: Field,
+                        fieldValue: Expression,
+                        violationsList: String): CodeBlock {
+        val type = field.declaringType
+        val violation = buildViolation(type, field, fieldValue)
+        return addViolation(violation, violationsList)
+    }
+
+    /**
+     * Constructs code which creates a [ConstraintViolation] of a composite validation rule and adds
+     * it to the given mutable [violationsList].
+     */
+    fun createCompositeViolation(type: TypeName, violationsList: String): CodeBlock {
+        val violation = buildViolation(type, null, null)
+        return addViolation(violation, violationsList)
+    }
+
+    private fun addViolation(violation: Expression, violationsList: String): CodeBlock =
+        CodeBlock
+            .builder()
+            .addStatement("\$N.add(\$L)", violationsList, violation)
+            .build()
+
+
+    private fun buildViolation(type: TypeName, field: Field?, fieldValue: Expression?): Expression {
+        var violationBuilder = ClassName(ConstraintViolation::class.java)
+            .newBuilder()
+            .chainSet("msg_format", LiteralString(value))
+            .chainSet("type_name", LiteralString(type.typeUrl()))
+        if (field != null) {
+            violationBuilder = violationBuilder.chainSet("field_path", pathFrom(field))
+        }
+        if (fieldValue != null) {
+            violationBuilder = violationBuilder.chainSet("field_value", fieldValue.packToAny())
+        }
+        return violationBuilder.chainBuild()
+    }
+
+    private fun pathFrom(field: Field): Expression {
```

Comment:
Isn't it `pathOf()`? I mean, a path from <some field> should probably a path to somewhere. Taken by someone who wants to get from this field to some particular place.